### DISCUSSION TITLE
Move the base point closer to the geometry

### DISCUSSION THREAD

**Discussion by santiago-vargas-de-kruijf** [mozman/ezdxf#925] [Q&A] [answered]
Hello, I have a question regarding Inserts and Blocks. During our development process, we've found some sample DXFs that have Insert entities which have a base_point away from the Insert center. This is what it looks like when I select the Insert in Autocad, the base point is far away from the center of the entity.

![image](https://github.com/mozman/ezdxf/assets/124168836/76bc3884-89d1-435b-8100-921ffe8e498c)

Using the approach provided by this article: https://www.autodesk.com/support/technical/article/caas/sfdcarticles/sfdcarticles/When-inserting-a-block-into-an-AutoCAD-drawing-the-base-point-of-the-block-is-far-away-of-the-block-geometry.html we are able to manually move the base point closer to the geometry, but we need a way to do this using Ezdxf.

Is this possible? 

When checking the Block of this specific Insert entity.

`insert.block().block.dxf.base_point`

The result is `(0,0,0)`, according to the docs https://ezdxf.readthedocs.io/en/stable/blocks/block.html?highlight=Block#ezdxf.entities.Block.dxf.base_point this is the default value.

**Comment by mozman** [comment, marked as answer]
Hi!

I looked at both solution and yes they can be replicated by `ezdxf`. I did not see that these solutions do update the block references itself but maybe this is done by AutoCAD automatically. If you move the base point, the existing block references will be positioned incorrectly.

All solution need the new `ezdxf` version v1.1.

Attached resources: 
- [TestBlock.zip](https://github.com/mozman/ezdxf/files/12616623/TestBlock.zip)
- [script.zip](https://github.com/mozman/ezdxf/files/12616761/script.zip)

The block "CIRCLE" has the base point at the origin (0, 0,  0):
![block_with_offset](https://github.com/mozman/ezdxf/assets/4230966/e0248214-dea7-4bd4-89e4-ccce27a687ba)

The first solution: moves all entities to the base point, existing block references are not updated. The new base point is defined by the center of the block bounding box (the center of the CIRCLE entity):

```Python
import pathlib
import ezdxf
from ezdxf import bbox, transform


CWD = pathlib.Path("~/Desktop/Outbox").expanduser()
if not CWD.exists():
    CWD = pathlib.Path(".")

doc = ezdxf.readfile(CWD/"TestBlock.dxf")

circle_block = doc.blocks.get("CIRCLE")
if not circle_block:
    print("Block CIRCLE not found")
    exit(1)

current_base_point = circle_block.base_point  # requires ezdxf v1.1

# get new base point from bounding box detection:
extents = bbox.extents(circle_block, fast=True)
new_base_point = extents.center  # or extents.extmin/extmax ...

offset = current_base_point - new_base_point
transform.translate(circle_block, offset)  # requires ezdxf v1.1

doc.saveas(CWD/"move_entities.dxf")
```

Because the block reference hasn't been updated the block reference moved to the origin:
![content_moved](https://github.com/mozman/ezdxf/assets/4230966/655f1070-309b-427e-8cd3-a43ef9983101)

The secong solution moves the base point of the block, existing block references are not updated:

```Python
# previous code required!
doc = ezdxf.readfile(CWD/"TestBlock.dxf")
circle_block = doc.blocks.get("CIRCLE")
circle_block.base_point = new_base_point

doc.saveas(CWD/"move_base_point.dxf")
```
Same result as for the first solution.

Updating the block references:

```Python
# previous code required!
# get ALL block references for block "CIRCLE"
block_refs = doc.query("INSERT[name=='CIRCLE']")
transform.translate(block_refs, -offset)

doc.saveas(CWD/"update_block_refs.dxf")
```

The block base point has been moved and the existing block references have been updated:
![updated_block_refs](https://github.com/mozman/ezdxf/assets/4230966/8e7791d2-9fb6-49d4-b71c-f17a235fbeaf)

## NOTES:
- Bounding box detection doesn't work for ACIS based entities and transformation also doesn't work for ACIS entities, so only the 2nd solution is applicable
- Dynamic blocks are not supported (undocumented feature)

**Reply by mozman** [reply]
Additional note: If you do not update the existing block references, all attached ATTRIB entities will remain in their original position because ATTRIB coordinates are not relative to the INSERT entity, but are relative to the origin of the parent layout (modelspace or block).

**Reply by santiago-vargas-de-kruijf** [reply]
Thank you for the detailed explanation and examples. This is really useful.
